pkg/resources/server_test.go

```diff
@@ -0,0 +1,327 @@
+			}
+
+			err := rs.Restart()
+			Expect(err).ToNot(HaveOccurred())
+		})
+	})
+	Context("Watch", func() {
+		var DepSock string
+		var fs utils.FakeFilesystem
+		var cleanTemp func()
+		BeforeSuite(func() {
+			DepSock = deprecatedSockDir
+		})
+		AfterSuite(func() {
+			deprecatedSockDir = DepSock
+		})
+		BeforeEach(func() {
+			fs = utils.FakeFilesystem{}
+			cleanTemp = fs.Use()
+			deprecatedSockDir = fs.RootDir
+		})
+		AfterEach(func() {
+			cleanTemp()
+		})
+		It("Watch socket and send notification of create successfully", func() {
+			grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("GetServer", mock.Anything).Return(grpcServer)
+			rsc.On("CreateServer").Return()
+			rsc.On("DeleteServer").Return()
+			rsc.On("Listen", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Serve", mock.Anything).Return()
+			rsc.On("Dial", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Stop").Return()
+			rsc.On("Close", mock.Anything).Return()
+
+			rs := resourceServer{
+				watchMode:   true,
+				rsConnector: rsc,
+				socketPath:  "fake",
+				stop:        make(chan interface{}),
+			}
+			go func() {
+				// Wait for watch to create the socket watcher
+				time.Sleep(5 * time.Millisecond)
+				event := fsnotify.Event{Name: "fake", Op: fsnotify.Create}
+				rs.socketWatcher.Events <- event
+			}()
+			err := rs.Watch()
+			Expect(err).ToNot(HaveOccurred())
+		})
+		It("Watch socket and send notification of create and failed to restart server", func() {
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("GetServer", mock.Anything).Return(nil)
+			rsc.On("CreateServer").Return()
+			rsc.On("Listen", mock.Anything, mock.Anything).Return(nil, errors.New("failed"))
+
+			rs := resourceServer{
+				watchMode:   true,
+				socketPath:  "fake",
+				rsConnector: rsc,
+				stop:        make(chan interface{}),
+			}
+			go func() {
+				// Wait for watch to create the socket watcher
+				time.Sleep(5 * time.Millisecond)
+				event := fsnotify.Event{Name: "fake", Op: fsnotify.Create}
+				rs.socketWatcher.Events <- event
+			}()
+			err := rs.Watch()
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(Equal("unable to restart server failed"))
+		})
+		It("Watch socket and getting failed error from watcher", func() {
+			rs := resourceServer{}
+			go func() {
+				// Wait for watch to create the socket watcher
+				time.Sleep(5 * time.Millisecond)
+				err := errors.New("failed")
+				rs.socketWatcher.Errors <- err
+			}()
+			err := rs.Watch()
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(Equal("inotify: failed"))
+		})
+		It("Watch socket then stop watcher", func() {
+			rs := resourceServer{stopWatcher: make(chan bool)}
+			go func() {
+				// Wait for watch to create the socket watcher
+				time.Sleep(5 * time.Millisecond)
+				rs.stopWatcher <- true
+			}()
+			err := rs.Watch()
+			Expect(err).ToNot(HaveOccurred())
```

Comment:
ensure watcher.Close() was called

Reply:
when send rs.stopWatcher <- true, then the Watch will call watcher.Close(), if that didn't happen the test will be stuck since it is waiting for return from Watch()